llama_index/llms/perplexity.py

```diff
@@ -0,0 +1,358 @@
+import json
+from typing import Any, Dict, Optional, Sequence
+import aiohttp
+import httpx
+import requests
+from llama_index.bridge.pydantic import Field
+from llama_index.callbacks import CallbackManager
+from llama_index.llms.base import (
+    LLM,
+    ChatMessage,
+    ChatResponse,
+    ChatResponseAsyncGen,
+    ChatResponseGen,
+    CompletionResponse,
+    CompletionResponseAsyncGen,
+    CompletionResponseGen,
+    LLMMetadata,
+    llm_chat_callback,
+    llm_completion_callback,
+)
+
+
+class Perplexity(LLM):
+    model: str = Field(description="The Perplexity model to use.")
+    temperature: float = Field(description="The temperature to use during generation.")
+    max_tokens: Optional[int] = Field(
+        default=None, description="The maximum number of tokens to generate."
+    )
+    api_key: str = Field(
+        default=None, description="The Perplexity API key.", exclude=True
+    )
+    api_base: str = Field(
+        default="https://api.perplexity.ai",
+        description="The base URL for Perplexity API.",
+    )
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description="Additional kwargs for the Perplexity API."
+    )
+    max_retries: int = Field(
+        default=10, description="The maximum number of API retries."
+    )
+    headers: Dict[str, str] = Field(
+        default_factory=dict, description="Headers for API requests."
+    )
+
+    def __init__(
+        self,
+        model: str = "mistral-7b-instruct",
+        temperature: float = 0.1,
+        max_tokens: Optional[int] = None,
+        api_key: Optional[str] = None,
+        api_base: Optional[str] = "https://api.perplexity.ai",
+        additional_kwargs: Optional[Dict[str, Any]] = None,
+        max_retries: int = 10,
+        callback_manager: Optional[CallbackManager] = None,
+        **kwargs: Any,
+    ) -> None:
+        additional_kwargs = additional_kwargs or {}
+        super().__init__(
+            model=model,
+            temperature=temperature,
+            max_tokens=max_tokens,
+            additional_kwargs=additional_kwargs,
+            max_retries=max_retries,
+            callback_manager=callback_manager,
+            api_key=api_key,
+            api_base=api_base,
+            **kwargs,
+        )
+        self.headers = {
+            "accept": "application/json",
+            "content-type": "application/json",
+            "authorization": f"Bearer {self.api_key}",
+        }
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "perplexity_llm"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        return LLMMetadata(
+            context_window=self._get_context_window(),
```

Comment:
I think it might be better to also let the user override this. It can be helpful in some cases to actually shrink it (for example, our default context window for OpenAI is 3900, to help avoid token overflow errors)

If the user does not provide it, then it could default back to one of these values

Reply:
I see, that makes sense. So the right way to do this would be to make context_window also another parameter provided by the user and set it up in init - and if it is None, we set it with _get_context_window()?

Reply:
Yup you got it 👍

Reply:
This is done, let me know if it is okay!